### ISSUE TITLE
[ENHANCEMENT] Add Log-Spectral Distance (LSD) metric for spatial frequency analysis

### ISSUE THREAD

**Issue by sohampatil01-svg** [mllam/neural-lam#506]
Currently, neural-lam relies on point-wise metrics (MSE, MAE, NLL, CRPS) which do not explicitly penalize the loss of high-frequency spatial information. In numerical weather prediction (NWP), this often leads to 'regression to the mean,' where models produce blurry fields that lose fine-scale structures like sharp fronts or small-scale variability.

I propose adding a  log_spectral_distance  metric to  metrics.py . This metric computes the distance between the power spectra of the prediction and target in the spatial frequency domain using 2D RFFT. It allows for a better assessment of whether a model preserves the correct energy distribution across different spatial scales.

### Key features:
- Uses  torch.fft.rfft2  for efficient computation.
- Differentiable, allowing it to be used as a training objective or validation metric.
- Supports  BaseRegularGridDatastore  by unstacking the  grid_index  into  (y, x)  if the grid shape is provided.

I have verified the mathematical implementation locally and it successfully distinguishes between blurred predictions and noisy predictions that have the same MSE.

**Comment by sohampatil01-svg** [comment]
@sadamov  @leifdenby @joeloskarsson @observingClouds   Could you please assign this issue to me so I can open the PR?   I've already implemented this feature and verified it with unit tests.

**Comment by observingClouds** [comment]
Hi @sohampatil01-svg,

Thanks for suggesting this improvement. May I challenge you and ask, if you can come up with a solution that works on irregular grids as well and not only on regularly gridded forecasts? I know this might be challenging, but it would make this enhancement available for a much larger set of forecasts.

**Comment by sohampatil01-svg** [comment]
"Thank you for the  challenge, @observingClouds. 
You are correct—standard FFT is a special case of a more general framework. To support irregular grids (like the icosahedral meshes used in global models), we can pivot from Euclidean Fourier transforms to Graph Signal Processing (GSP).

  In GSP, 'spatial frequency' is defined by the eigenvalues of the Normalized Graph Laplacian:

<img width="402" height="106" alt="Image" src="https://github.com/user-attachments/assets/3611e5a4-3a69-46b1-8c5f-4ee3da23ca92" />

  $A$ = adjacency matrix and $D$ = degree matrix. 
The 'high-frequency' content of a forecast $x$ is captured by the Graph Dirichlet Energy:
 
<img width="542" height="122" alt="Image" src="https://github.com/user-attachments/assets/c30f9c7d-8a99-4d18-a1a2-ec9b2c6eceee" />

  To implement an Unstructured Log-Spectral Distance, we can compare the energy distribution across different 'eigen-bands.' Since explicit eigendecomposition is $O(N^3)$, I propose using a Stochastic Trace Estimator (Hutchinson’s trick) to
approximate the Power Spectral Density (PSD) through Laplacian moments:

<img width="572" height="126" alt="Image" src="https://github.com/user-attachments/assets/73f81d2f-d603-4216-a60d-bfe253ab4106" />

  Here, $L^k$ acts as a high-pass filter that extracts energy at increasingly fine spatial scales. This would make the metric natively compatible with any graph structure in neural-lam, including the hierarchical levels.

  I’ll start researching the most efficient way to compute these power-trace estimators in PyTorch to ensure it remains performant for large global graphs. 

Does this GSP-based approach align with the level of generalization you were looking for?"

**Comment by sohampatil01-svg** [comment]
Hii @observingClouds  !!!...
I have implemented the Challlenge ! 
Generalized the Log-Spectral Distance (LSD) implementation to support irregular/unstructured grids as well as regular ones.
Please have a look at  it !  if any thing needs to be changed or wrong please let me know !!...
please check the pr 508 !


What has been implemented:
To move beyond 2D FFTs (which are limited to regular grids), I implemented a Graph Signal Processing (GSP) approach for the LSD metric:

1. Laplacian Moments ($m_k = x^T L^k x$): On unstructured meshes, the metric now computes spectral moments using the
Normalized Graph Laplacian. These moments act as spectral estimators that capture energy distribution across different 'eigen-bands' of the graph, providing a mathematically sound analog to Fourier frequency bins for arbitrary topologies.

2. Hybrid Detection: The log_spectral_distance function now automatically detects the grid type. It uses fast 2D RFFT for regular grids and falls back to Sparse Laplacian Power Iterations for unstructured ones.
   
3. Model Integration: I updated ARModel and BaseGraphModel to automatically extract the mesh connectivity (edge_index) and pass it to the loss function. This ensures that the LSD metric is 'topology-aware' for any LAM configuration (1level, multiscale, or hierarchical).
   
4. Verification: I've added tests for the GSP-based branch and verified that existing regular grid tests still pass. I also updated the CHANGELOG.md to reflect these additions.
